### DISCUSSION TITLE
Running Spawn inside a Docker container

### DISCUSSION THREAD

**Discussion by AntoineGautier** [lbl-srg/modelica-buildings#3969] [Q&A] [answered]
When I run Spawn locally from `/home/reituag/gitrepo/modelica-buildings` I get the following JSON intermediary file that is then used to create the FMU:
```json
{
  "version": "0.1",
  "EnergyPlus": {
    "idf": "/home/reituag/gitrepo/modelica-buildings/Buildings/Resources/Data/ThermalZones/EnergyPlus_9_6_0/Examples/SingleFamilyHouse_TwoSpeed_ZoneAirBalance/SingleFamilyHouse_TwoSpeed_ZoneAirBalance.idf",
    "weather": "/home/reituag/gitrepo/modelica-buildings/Buildings/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw",
```

However, when I run Spawn from `/mnt/shared/modelica-buildings` inside a Docker container, I get:

```json
{
  "version": "0.1",
  "EnergyPlus": {
    "idf": "/tmp/JModelica.org/jm_tmpzh05d4xs/resources/0/SingleFamilyHouse_TwoSpeed_ZoneAirBalance.idf",
    "weather": "/tmp/JModelica.org/jm_tmpzh05d4xs/resources/1/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw",
```

and the following error:
```
Executing '""/mnt/home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/bin/spawn-0.5.0-c10e8c6d7e"" energyplus create-fmu  --output-path ""/mnt/shared/modelica-buildings/spawn-AirHeating/EnergyPlus.fmu"" ""/mnt/shared/modelica-buildings/spawn-AirHeating/ModelicaBuildingsEnergyPlus.json""' failed to generate fmu '/mnt/shared/modelica-buildings/spawn-AirHeating/EnergyPlus.fmu'
```

So perhaps the question is how the JSON properties `idf` and `weather` are assigned and whether a specific setting is missing in my container.

Thanks for any insight.

**Comment by mwetter** [comment]
The error message looks strange to me. It contains
```
Executing '""/mnt/home/reituag/spawn/...
```
and
```
"/mnt/shared/modelica-buildings/spawn-AirHeating/EnergyPlus.fmu"
```
in the same message. This looks like your docker has spawn installed in `/mnt/home/reituag` (which you perhaps export to the docker as otherwise the executable would not know about its existence) and this is on the PATH before it picks up the spawn binary from `/mnt/shared/modelica-buildings/...`.

The C code that invokes spawn searches for the spawn binary in various places, see
https://github.com/lbl-srg/modelica-buildings/blob/8584e8030b602fec7da07b682bf8dd750e57e5e3/Buildings/Resources/src/ThermalZones/EnergyPlus_9_6_0/C-Sources/BuildingInstantiate.c#L383

Eventually, as the path to the Modelica Buildings Library is exported to the docker, and spawn knows the top-level directory name, it will find the binary of the Modelica Buildings Library. However, in your error message, it looks like spawn first finds it in `/mnt/home/reituag...`.

**Comment by AntoineGautier** [comment, marked as answer]
>  This looks like your docker has spawn installed in /mnt/home/reituag (which you perhaps export to the docker as otherwise the executable would not know about its existence) and this is on the PATH before it picks up the spawn binary from /mnt/shared/modelica-buildings/...

Yes, I bind mount the directory with the Spawn executable, that I manually installed outside the MBL directory on the host.
This is because I am using the development branch of MBL and not the public release which comes with a bundled Spawn executable. 
The `PATH` environment variable of the container is set to include the path where the Spawn directory is mounted in the container. 
So, from within the container:
```sh
> which spawn
/mnt/home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/bin/spawn
```

But I found the solution.
I needed to mount the whole Spawn installation directory, not only the `bin` subdirectory with the Spawn executable as I previously did (the one that is included in the `PATH`). This is because `epfmi.so` is required at runtime and:
```sh
> tree ~/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/
/home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/
├── bin
│   ├── spawn -> spawn-0.5.0-c10e8c6d7e
│   └── spawn-0.5.0-c10e8c6d7e
├── etc
│   └── Energy+.idd
├── lib
│   └── epfmi.so
└── README.md
``` 
So, replacing:
```sh
docker run -v /home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/bin:/mnt/home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux/bin:ro ...
```
with
```sh
docker run -v /home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux:/mnt/home/reituag/spawn/Spawn-light-0.5.0-c10e8c6d7e-Linux:ro
```
solved the issue for me.

Note that the error message that I shared in my previous post was taken from the OCT log file. 
The error message in the terminal generated by the Spawn executable clearly pointed to the issue with the `epfmi.so` path.
